src/lib/rmt/base.rb

```diff
@@ -0,0 +1,73 @@
+# Copyright (c) 2018 SUSE LLC.
+#  All Rights Reserved.
+
+#  This program is free software; you can redistribute it and/or
+#  modify it under the terms of version 2 or 3 of the GNU General
+#  Public License as published by the Free Software Foundation.
+
+#  This program is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.   See the
+#  GNU General Public License for more details.
+
+#  You should have received a copy of the GNU General Public License
+#  along with this program; if not, contact SUSE LLC.
+
+#  To contact SUSE about this file by physical or electronic mail,
+#  you may find current contact information at www.suse.com
+
+require 'yaml'
+
+Yast.import 'Report'
+
+module RMT
+end
+
+class RMT::Base < Yast::Client
+  include Yast::UIShortcuts
+  include Yast::Logger
+  include Yast::I18n
+
+  CONFIG_FILENAME = '/etc/rmt.conf'.freeze
+
+  def self.read_config_file
+    begin
+      data = Yast::SCR.Read(Yast.path('.target.string'), CONFIG_FILENAME)
+      config = YAML.safe_load(data)
+    rescue StandardError => e
+      log.warn 'Reading config file failed: ' + e.to_s
+    end
+
+    config ||= {}
+    config['scc'] ||= {}
+    config['scc']['username'] ||= ''
+    config['scc']['password'] ||= ''
+
+    config['database'] ||= {}
+    config['database']['database'] ||= 'rmt'
+    config['database']['username'] ||= 'rmt'
+    config['database']['password'] ||= ''
+    config['database']['hostname'] ||= 'localhost'
```

Comment:
just tip. This defaults is a bit unreadable. I suggest this simple code to merge default to hash ( not handling false values and array, but if you are interested in more complex and fully working with all values see https://stackoverflow.com/a/9381776 )

```
loaded = {
  "scc" => { "username" => "test" }
}

DEFAULTS = {
  "scc" => {
    "username" => "",
    "password" => ""
  },
  "database" => {
     "database" => "rmt",
     "username" => "rmt"
     # ...
  }
}.freeze

simple_deep_merger = proc { |key, v1, v2| v1.is_a?(Hash) ? v1.merge(v2, &simple_deep_merger) : v1 ? v1 : v2 }
puts DEFAULTS.merge(loaded, &simple_deep_merger).inspect
```

Reply:
Proc's hacky, but makes sense. I'd maybe create a private method for merging.